modular_skyrat/master_files/code/modules/client/preferences/middleware/food.dm

```diff
@@ -0,0 +1,163 @@
+					continue
+
+				if(preferences.food[food_entry] == "[FOOD_LIKED]")
+					liked_food_length++
+					if(liked_food_length > MAXIMUM_LIKES)
+						preferences.food.Remove(food_entry)
+
+			if(food_flag == "[FOOD_LIKED]" ? liked_food_length >= MAXIMUM_LIKES : liked_food_length > MAXIMUM_LIKES) // Equals as well, if we're setting a liked food!
+				return TRUE // Fuck you, look your mistakes in the eye.
+
+			preferences.food[food_name] = food_flag
+			return TRUE
+
+/datum/preferences/ui_state(mob/user)
+	return GLOB.always_state
+
+/datum/food_prefs_menu/ui_status(mob/user, datum/ui_state/state)
+	return UI_INTERACTIVE // Prefs can be accessed from anywhere.
+
+/datum/food_prefs_menu/ui_static_data(mob/user)
+	return list(
+		"food_types" = GLOB.food_ic_flag_to_point_values,
+	)
+
+/datum/food_prefs_menu/ui_data(mob/user)
+	var/datum/preferences/preferences = user.client.prefs
+
+	var/datum/species/species = preferences.read_preference(/datum/preference/choiced/species)
+
+	return list(
+		"selection" = preferences.food,
+		"points" = calculate_points(preferences),
+		"enabled" = preferences.food["enabled"],
+		"invalid" = is_food_invalid(preferences),
+		"pref_literally_does_nothing" = !initial(species.allow_food_preferences),
+	)
+
+/// Checks the provided preferences datum to make sure the food pref values are valid. Does not check if the food preferences value is null.
+/datum/food_prefs_menu/proc/is_food_invalid(datum/preferences/preferences)
+	var/liked_food_length = 0
+	var/disliked_food_length = 0
+	var/toxic_food_length = 0
+
+	for(var/food_entry in GLOB.food_ic_flag_to_point_values)
+		var/list/food_points_entry = GLOB.food_ic_flag_to_point_values[food_entry]
+		var/food_preference = preferences.food[food_entry] || food_points_entry["[FOOD_DEFAULT]"]
+
+		if(food_points_entry["[FOOD_OBSCURE]"])
+			continue
+
+		switch(food_preference) // Doesn't use defines cause I need this in string form.
+			if("1")
+				liked_food_length++
+			if("2")
+				disliked_food_length++
+			if("3")
+				toxic_food_length++
+
+	if(liked_food_length > MAXIMUM_LIKES)
+		return "too many like choices"
+	if(disliked_food_length < MINIMUM_REQUIRED_DISLIKES)
+		return "too few dislike choices"
+	if(toxic_food_length < MINIMUM_REQUIRED_TOXICS)
+		return "too few toxic choices"
+	if(calculate_points(preferences) < 0)
+		return "not enough points"
+
+/// Calculates the deviance points for food.
+/datum/food_prefs_menu/proc/calculate_points(datum/preferences/preferences)
+	var/points = 0
+
+	for(var/food_entry in preferences.food)
+		var/list/food_flag = preferences.food[food_entry]
+		var/list/food_points_entry = GLOB.food_ic_flag_to_point_values[food_entry]
+		if(!food_points_entry)
+			continue
+
+		var/default_food_flag = food_points_entry["[FOOD_DEFAULT]"]
+		var/point_value = food_points_entry[food_flag]
+
+		if(point_value && (GLOB.food_flag_to_order_value[food_flag] > GLOB.food_flag_to_order_value[default_food_flag]))
```

Comment:
This feels like a REALLY strange way to do it. Should just use negative values when it's below the default "preference" for the food.